File: dynamic_programming/fibonacci.rb
def fib_rec(n, memo = {})
  # memoized ruby implementation (scope)
  return memo[n] if memo[n]
  return 1 if n <= 2

  memo[n-2] = fib_rec(n-2, memo)
  memo[n-1] = fib_rec(n-1, memo)
  memo[n] = memo[n-2] + memo[n-1]
end

def fib_btm_up(n)
  # linear constant space bottom up implementation
  last1 = 0
  last2 = 0
  f = 0

  1.upto(n) do |k|
    f = k <= 2 ? 1 : last1 + last2
    last1 = last2
    last2 = f
  end
  return f
end

File: practice/fibonacci.js
// bottom up implementation
// linear time
// constant space

function btmUpFibonacci(n) {
  var idx = 1;
  var last1 = 1;
  var last2 = 1;
  while (idx <= n) {
    var f = idx <= 2 ? 1 : last1 + last2;
    last2 = last1;
    last1 = f;
    idx += 1;
  }
  return f;
}

// memoized version
var _memo = {};

function memoFibonacci(n) {
  if (_memo[n])
    return _memo[n];
  if (n <= 2)
    return 1;

  _memo[n - 2] = memoFibonacci(n - 2);
  _memo[n - 1] = memoFibonacci(n - 1);
  return _memo[n - 2] + _memo[n - 1];
}
